token/core/zkatdlog/nogh/v1/crypto/upgrade/service.go

```diff
@@ -0,0 +1,249 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package upgrade
+
+import (
+	"bytes"
+	"crypto/rand"
+	"slices"
+
+	"github.com/hyperledger-labs/fabric-smart-client/pkg/utils/errors"
+	v1 "github.com/hyperledger-labs/fabric-token-sdk/token/core/fabtoken/v1"
+	token2 "github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/nogh/v1/crypto/token"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/logging"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/token"
+)
+
+const (
+	ChallengeSize = 32
+)
+
+type (
+	Signature = []byte
+)
+
+//go:generate counterfeiter -o mock/des.go -fake-name Deserializer . Deserializer
+
+type Deserializer interface {
+	GetOwnerVerifier(id driver.Identity) (driver.Verifier, error)
+}
+
+//go:generate counterfeiter -o mock/ip.go -fake-name IdentityProvider . IdentityProvider
+
+type IdentityProvider interface {
+	GetSigner(id driver.Identity) (driver.Signer, error)
+}
+
+type Service struct {
+	Logger                          logging.Logger
```

Comment:
We can keep these private and when we want to extend them we can make the ones we are interested in public.